source/MaterialXRender/ShaderRenderer.h

```diff
@@ -110,6 +110,9 @@ class MX_RENDER_API ShaderRenderer
     /// Validate inputs for the program.
     virtual void validateInputs() { }
 
+    /// Update the program with value of the uniform.
+    virtual void updateUniform(ShaderPort* uniform) = 0;
```

Comment:
I'd recommend giving this virtual method a default definition, so that we don't change this into an abstract base class (which has implications for the complexity of Python bindings).

Reply:
If we're doing this. Can we throw the "unimplemented" exception here? This will cover all derived classes instead of just `OSLRenderer`. Then you can get rid of the override there.

Reply:
I would suggest for generality to pass over a `ValuePtr`. Then the update can be done w/o adding in the `ShaderPort` dependence.